google/cloud/opentelemetry/resource_detector.cc

```diff
@@ -0,0 +1,241 @@
+  SetEnvAttribute(attributes, sc::kFaasName, "GAE_SERVICE");
+  SetEnvAttribute(attributes, sc::kFaasVersion, "GAE_VERSION");
+  SetEnvAttribute(attributes, sc::kFaasInstance, "GAE_INSTANCE");
+
+  auto zone = Tail(MD(json, {"instance", "zone"}));
+  SetAttribute(attributes, sc::kCloudAvailabilityZone, zone);
+  auto const pos = zone.rfind('-');
+  SetAttribute(attributes, sc::kCloudRegion, zone.substr(0, pos));
+}
+
+void Gce(opentelemetry::sdk::resource::ResourceAttributes& attributes,
+         nlohmann::json const& json) {
+  attributes.SetAttribute(sc::kCloudPlatform, "gcp_compute_engine");
+  SetAttribute(attributes, sc::kHostType,
+               Tail(MD(json, {"instance", "machineType"})));
+  SetAttribute(attributes, sc::kHostId, MD(json, {"instance", "id"}));
+  SetAttribute(attributes, sc::kHostName, MD(json, {"instance", "name"}));
+
+  auto zone = Tail(MD(json, {"instance", "zone"}));
+  attributes.SetAttribute(sc::kCloudAvailabilityZone, zone);
+  auto const pos = zone.rfind('-');
+  attributes.SetAttribute(sc::kCloudRegion, zone.substr(0, pos));
+}
+
+class ResourceDetector final
+    : public opentelemetry::sdk::resource::ResourceDetector {
+ public:
+  explicit ResourceDetector(otel_internal::HttpClientFactory factory,
+                            Options options)
+      : client_factory_(std::move(factory)), options_(std::move(options)) {}
+
+  opentelemetry::sdk::resource::Resource Detect() override {
+    auto client = client_factory_(options_);
+    rest_internal::RestRequest request;
+    request.SetPath(absl::StrCat(internal::GceMetadataScheme(), "://",
+                                 internal::GceMetadataHostname(),
+                                 "/computeMetadata/v1/"));
+    request.AddHeader("metadata-flavor", "Google");
+    request.AddQueryParameter("recursive", "true");
+    rest_internal::RestContext context;
+    auto response = client->Get(context, request);
+
+    auto log_error = [](Status const& status) {
+      GCP_LOG(INFO) << "Could not query the metadata server. status=" << status;
+      return opentelemetry::sdk::resource::Resource::GetEmpty();
+    };
+
+    if (!response) {
+      auto status = std::move(response).status();
+      if (status.code() == StatusCode::kUnavailable) {
+        // This is most likely a "curl: (6) Could not resolve host" error, which
+        // means the application is not running on GCP.
+        return opentelemetry::sdk::resource::Resource::GetEmpty();
+      }
+      return log_error(status);
+    }
+    if (!response) return log_error(std::move(response).status());
+    if (rest_internal::IsHttpError(**response)) {
+      auto status = rest_internal::AsStatus(std::move(**response));
+      return log_error(status);
+    }
+
+    auto payload =
+        rest_internal::ReadAll(std::move(**response).ExtractPayload());
+    if (!payload.ok()) return log_error(payload.status());
+    auto json = nlohmann::json::parse(*payload, nullptr, false);
```

Comment:
You need more validation.  First, we should check that the headers returned matched our expectations.  Second, some people have ISPs that return HTML (full of adds) when you send a HTTP request to an unknown hostname.